From 81c55c192aeabafbd80e9dab12b899bbe262d4b5 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:04:54 -0400
Subject: [usb] usbtest: Add timetout to simple_io()

Message-id: <1410887463-20674-223-git-send-email-dzickus@redhat.com>
Patchwork-id: 95174
O-Subject: [RHEL7 PATCH 222/591] usb: usbtest: Add timetout to simple_io()
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit e5e4746510d140261918aecce2e5e3aa4456f7e9
Author: Roger Quadros <rogerq@ti.com>
Date:   Wed Dec 18 15:40:10 2013 +0530

    usb: usbtest: Add timetout to simple_io()

    Without a timetout some tests e.g. test_halt() can remain stuck forever.

    Signed-off-by: Roger Quadros <rogerq@ti.com>
    Reviewed-by: Felipe Balbi <balbi@ti.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/misc/usbtest.c b/drivers/usb/misc/usbtest.c
index 446ff55..a782bfc 100644
--- a/drivers/usb/misc/usbtest.c
+++ b/drivers/usb/misc/usbtest.c
@@ -10,6 +10,7 @@
 
 #include <linux/usb.h>
 
+#define SIMPLE_IO_TIMEOUT 10000 /* in milliseconds */
 
 /*-------------------------------------------------------------------------*/
 
@@ -366,6 +367,7 @@ static int simple_io(
  int   max = urb->transfer_buffer_length;
  struct completion completion;
  int   retval = 0;
+ unsigned long  expire;
 
  urb->context = &completion;
  while (retval == 0 && iterations-- > 0) {
@@ -378,9 +380,15 @@ static int simple_io(
   if (retval != 0)
    break;
 
-  /* NOTE:  no timeouts; can't be broken out of by interrupt */
-  wait_for_completion(&completion);
-  retval = urb->status;
+  expire = msecs_to_jiffies(SIMPLE_IO_TIMEOUT);
+  if (!wait_for_completion_timeout(&completion, expire)) {
+   usb_kill_urb(urb);
+   retval = (urb->status == -ENOENT ?
+      -ETIMEDOUT : urb->status);
+  } else {
+   retval = urb->status;
+  }
+
   urb->dev = udev;
   if (retval == 0 && usb_pipein(urb->pipe))
    retval = simple_check_buf(tdev, urb);
-- 
1.7.1